wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileSchemaCachedProvider.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.devfile.server;
+
+import static org.eclipse.che.api.devfile.server.Constants.SCHEMA_LOCATION;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.ref.SoftReference;
+import java.util.stream.Collectors;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+
+/** Loads a schema content and stores it in soft reference. */
+@Singleton
+public class DevfileSchemaCachedProvider {
```

Comment:
What if we do like this ?
```java
import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.io.Resources;
import com.google.inject.Singleton;

import java.io.IOException;
import java.lang.ref.SoftReference;
import java.nio.charset.Charset;

@Singleton
public class DefFileSchemaProvider {
  private final String SCHEMA_LOCATION = "";

  private SoftReference<String> schemaRef = new SoftReference(null);

  public String getString() throws IOException {
    String schema = schemaRef.get();
    if (schema == null) {
      synchronized (this) {
        if (schema == null) {
          schema =
              Resources.asCharSource(
                      Resources.getResource(SCHEMA_LOCATION), Charset.defaultCharset())
                  .read();
          schemaRef = new SoftReference(schema);
        }
      }
    }
    return schema;
  }

  public JsonNode getJsoneNode() throws IOException {
    return JsonLoader.fromString(getString());
  }
}
```

Reply:
1. this is the strangest way to load file i've ever saw :)
2. don;t think double-checked-locking is needed here. there is no danger to load wrong/different data for the threads;
3. agree that this provider shold be used for validation also, nice idea.